keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,1096 @@
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+Kubelet should not fail when during image pulling an event publishing to Pod fails, this should
+be ignored by kubelet and / or retried later if events are queued to be delivered. Queueing such
+events is out of scope of this proposal.
+
+###### What are other known failure modes?
+
+Nothing should fail, if the feature is enabled, but not working (either kubelet or runtime does
+not suppor proposed call) - the the fallback happens to the old functionality.
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
```

Comment:
Possible alternative: publish image pulling notification via a (Kubernetes API) subresource, and only when the API server informs the kubelet that someone is listening.

Reply:
> How would that someone subscribe for subresource updates?

At the API level, this could be a **watch**-style GET that requests only the status for an object.

> Do I understand right that this would only show real-time progress, without historical data?

We could define and design it either way.

> Say, if I deployed a Pod, and I check the Pod after 10 minutes, only then I would get the latest progress figures without knowing how fast if was doing previous 10 minutes?

We could define and design it either way.
- maybe the kubelet remembers some time-series data about the image pull (which I think makes sense, because images can belong to several Pods)
- maybe it goes into the API inside the Node object (doesn't sound ideal)
- maybe there's a new PodFetchMetrics object (sounds worse!)
- maybe nothing remembers metrics and we only have the start time and amount fetched so far

> Is it an existing concept that API server can inform that particular subresource is being listened upon? 

The API server might do an HTTP GET to a new path that the kubelet serves, and repeat that once a second _only whilst the watch happens_. And actual watches are already a thing. So mostly not new.

If we want the kubelet to push updates, that would need more novelty.

---

To be clear, documenting that there is an alternative that we've considered is not the same as selecting the alternative. It does however help show our thinking.

Reply:
At the CRI level, what I'm imagining is still `ImagePull` but with updates. The difference is around different ways to report progress to the API server. Examples:
- writes to the `status` subresource of any Pod waiting on the image pull
- create an Event at intervals whenever a Pod is waiting on the image pull
- _streaming_ updates, but only when someone fetches the `imagepullprogress` subresource of a relevant Pod
- _streaming_ updates, but only when someone fetches the `imagepullprogress` subresource of a relevant Node

&hellip;and more!

Reply:
Postponing kubelet implementation of events to beta stage.

Reply:
Thanks for the ideas @sftim !